### PR TITLE
improve some error messages.

### PR DESCRIPTION
**Pull request by brendandburns**
Fixes https://github.com/kubernetes/kubernetes/issues/9944

@pwittrock @deads2k 
@kubernetes/kubectl

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31675)

### CODE REVIEW

File: pkg/kubectl/cmd/exec.go [hunk 1]
```diff
@@ -48,6 +48,10 @@ var (
 		kubectl exec 123456-7890 -c ruby-container -i -t -- bash -il`)
 )
 
+const (
+	execUsageStr = "expected 'exec <pod-name> <command> [arg1] [arg2] ... [argN]'.\n<pod-name> and <command> are required parameter for the exec command"
```
Comment (jwforres) [inline pkg/kubectl/cmd/exec.go]: Don't we normally refer to them as arguments not parameters ?

@fabianofranz

Comment (smarterclayton) [inline pkg/kubectl/cmd/exec.go]: We generally use UPPERCASE argument names pretty consistently across all the commands.  Random sample of kubectl shows we use it for values to define the type everywhere in usage.  I think we should probably be consistent here.

Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: DONE

Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: :P

**Comment by brendandburns** [comment]
Comments addressed, unit test fixed.  PTAL.

Thanks
--brendan

File: pkg/kubectl/cmd/exec.go [hunk 2]
```diff
@@ -48,6 +48,10 @@ var (
 		kubectl exec 123456-7890 -c ruby-container -i -t -- bash -il`)
 )
 
+const (
+	execUsageStr = "expected 'exec POD_NAME COMMAND [ARG1] [ARG2] ... [ARGN]'.\nPOD_NAME and COMMAND are required arguments for the exec command"
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/exec.go]: Is this correct? `exec` takes `--` to start the `COMMAND` section. We should make it consistent with the usage, like 

```expected 'exec POD -- COMMAND [args...]'.\nPOD and COMMAND are required ...``` 

or even use the actual usage

```
execUsageStr = "expected %q.\nPOD and COMMAND are required arguments for the exec command."

(...)

return cmdutil.UsageError(cmd, execUsageStr, cmd.Use)
```

File: pkg/kubectl/cmd/exec.go [hunk 3]
```diff
@@ -135,19 +139,19 @@ type ExecOptions struct {
 func (p *ExecOptions) Complete(f *cmdutil.Factory, cmd *cobra.Command, argsIn []string, argsLenAtDash int) error {
 	// Let kubectl exec follow rules for `--`, see #13004 issue
 	if len(p.PodName) == 0 && (len(argsIn) == 0 || argsLenAtDash == 0) {
-		return cmdutil.UsageError(cmd, "POD is required for exec")
+		return cmdutil.UsageError(cmd, execUsageStr)
 	}
 	if len(p.PodName) != 0 {
-		printDeprecationWarning("exec POD", "-p POD")
+		printDeprecationWarning("exec <pod-name>", "-p <pod-name>")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/exec.go]: Keep using uppercase for arguments for consistency: `printDeprecationWarning("exec POD", "-p POD")`. Also `<...>` denotes an optional argument, which is not the case here.

File: pkg/kubectl/cmd/logs.go [hunk 4]
```diff
@@ -53,6 +53,10 @@ var (
 		kubectl logs --since=1h nginx`)
 )
 
+const (
+	logsUsageStr = "expected 'exec POD [CONTAINER_NAME]'.\nPOD_NAME is a required argument for the logs command"
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/logs.go]: Same as the comment above.

File: pkg/kubectl/cmd/scale.go [hunk 5]
```diff
@@ -120,10 +115,18 @@ func RunScale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
...
+	if resource.IsUsageError(err) {
+		return cmdutil.UsageError(cmd, err.Error())
+	}
 	if err != nil {
 		return err
 	}
 
+	count := cmdutil.GetFlagInt(cmd, "replicas")
+	if count < 0 {
+		return cmdutil.UsageError(cmd, "The --replicas=<count> flag is required, and <count> must be greater than or equal to 0")
```
Comment (fabianofranz) [inline pkg/kubectl/cmd/scale.go]: `--replicas=COUNT` or `--replicas=count` since the value is not optional.

Comment (fabianofranz) [inline pkg/kubectl/cmd/scale.go]: /me must add relevant POSIX documentation links to [kubectl-conventions.md](https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md).

https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap12.html

File: pkg/kubectl/cmd/exec.go (hunk 3 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: done.  (though I think `[foo]` denotes optional whilst `<foo>` denotes required.

File: pkg/kubectl/cmd/logs.go (hunk 4 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/logs.go]: done.

File: pkg/kubectl/cmd/scale.go (hunk 5 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/scale.go]: done.

File: pkg/kubectl/cmd/exec.go (hunk 2 above)
Comment (brendandburns) [inline pkg/kubectl/cmd/exec.go]: That's an option, but it's not required, it's only needed if you want to pass flags to the command that you execute `kubectl exec foo sh` works just fine.

**Comment by brendandburns** [comment]
Comments addressed, please re-check.

Thanks!
--brendan

**Comment by fabianofranz** [comment]
LGTM

**Comment by fabianofranz** [comment]
@smarterclayton mind tagging?

**Comment by deads2k** [comment]
> @smarterclayton mind tagging?

Done.  Could use a squash if you catch it before the bot does.

File: pkg/kubectl/cmd/logs.go [hunk 6]
```diff
@@ -53,6 +53,10 @@ var (
 		kubectl logs --since=1h nginx`)
 )
 
+const (
+	logsUsageStr = "expected 'logs POD_NAME [CONTAINER_NAME]'.\nPOD_NAME is a required argument for the logs command"
```
Comment (kargakis) [inline pkg/kubectl/cmd/logs.go]: logs support also replica sets and replication controllers

**Pull request merged**